### PR TITLE
feat(payments_v2): add payment_confirm_intent api endpoint

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This PR introduces the `payment-confirm-intent` endpoint. This is used to confirm the payment intent with payment method data.

This PR also includes changes to the payment attempt for api v2. Few columns have been added and few have been removed. The major changes that have been made to payment attempt are

- The id type is changed to GlobalId
- Amount related details are stored in a struct in the domain model. This is to ensure that all the operations / getters / setters related to amount are available in a single struct.
- Few fields that have been removed as a part of backwards compatibility removal have been deleted in the database.
- The address details have been de - normalized and stored in the payment attempt table. This is because we will not be reusing the address.

This PR contains only the minimum changes that are required to call the connector. 
- There is a bare minimum implementation of routing which is based on the fallback connector config of the business profile.
- There is no surcharge, decision manager, tokenization and other features. These would be refactored and slowly added to the v2 feature set.
- There is no implementation to save the card. These features would eventually be added to v2.

### Additional Changes

- [x] This PR modifies the API contract
         Add new api endpoint to payments. The openapi reference has been generated. 
Screenshot 2024-10-11 at 11 00 15 AM

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

- Create the payment intent
- Confirm the payment intent.
```bash
curl --location 'http://localhost:8080/v2/payments/12345_pay_0192b430fb0272b2a4ef5efdbfb06f68/confirm-intent' \
--header 'x-profile-id: pro_an7sjopN2AXVvhEEYvSm' \
--header 'x-client-secret: 12345_pay_0192b430fb0272b2a4ef5efdbfb06f68_secret_0192b430fb0272b2a4ef5f08095c2b1b' \
--header 'Content-Type: application/json' \
--header 'api-key: pk_dev_78d1c87dc958419587490ece50cbfd0c' \
--data '{
    "payment_method_data": {
            "card": {
                "card_number": "4242424242424242",
                "card_exp_month": "01",
                "card_exp_year": "27",
                "card_holder_name": "John Doe",
                "card_cvc": "100"
            }
    },
    "payment_method_type": "card",
    "payment_method_subtype": "credit"
}'
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code

### CODE REVIEW

File: crates/router/src/core/payments/transformers.rs [hunk 1]
```diff
@@ -140,19 +163,212 @@ where
...
+        access_token: None,
+        session_token: None,
+        reference_id: None,
+        payment_method_status: None,
+        payment_method_token: None,
+        connector_customer: None,
+        recurring_mandate_payment_data: None,
+        // TODO: This has to be generated as the reference id based on the connector configuration
+        // Some connectros might not accept accept the global id. This has to be done when generating the reference id
+        connector_request_reference_id: payment_data
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/transformers.rs:330]: we should give priority to merchant_reference_id right?

File: crates/router/src/core/payments/transformers.rs [hunk 2]
```diff
@@ -592,6 +822,72 @@ where
...
+
+        let amount = api_models::payments::ConfirmIntentAmountDetailsResponse::foreign_from((
+            &payment_intent.amount_details,
+            &payment_attempt.amount_details,
+        ));
+
+        let connector = payment_attempt
+            .connector
+            .clone()
+            .get_required_value("connector")
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/transformers.rs:854]: can we throw a 5xx here instead?

File: crates/hyperswitch_domain_models/src/payments/payment_attempt.rs [hunk 3]
```diff
@@ -266,6 +340,80 @@ impl PaymentAttempt {
...
+        let id = id_type::GlobalAttemptId::generate(&cell_id);
+        let intent_amount_details = payment_intent.amount_details.clone();
+
+        // TODO: move this to a impl function on payment attempt
+        let attempt_amount_details = AttemptAmountDetails {
+            net_amount: intent_amount_details.order_amount,
+            amount_to_capture: None,
+            surcharge_amount: None,
+            tax_on_surcharge: None,
+            amount_capturable: MinorUnit::new(0),
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/hyperswitch_domain_models/src/payments/payment_attempt.rs:362]: we should not be using the default values for amount, I would prefer to be implicit and provide the value zero.

File: crates/router/src/core/payments/operations/payment_confirm_intent.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                payment_intent_update,
+                key_store,
+                storage_scheme,
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Unable to update payment intent")?;
+        payment_data.payment_intent = updated_payment_intent;
+
+        let current_payment_attempt = payment_data.payment_attempt.clone();
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/operations/payment_confirm_intent.rs:303]: This has already been discussed, The refactor will be taken up in subsequent PRs

File: crates/api_models/src/payments.rs [hunk 5]
```diff
@@ -4396,6 +4471,124 @@ pub struct PaymentsResponse {
...
+    /// The payment method type to be used for the payment. This should match with the `payment_method_data` provided
+    #[schema(value_type = Option<PaymentMethod>, example = "card")]
+    pub payment_method_type: api_enums::PaymentMethod,
+
+    /// The payment method subtype to be used for the payment. This should match with the `payment_method_data` provided
+    #[schema(value_type = PaymentMethodType, example = "apple_pay")]
+    pub payment_method_subtype: api_enums::PaymentMethodType,
+
+    /// The shipping address for the payment. This will override the shipping address provided in the create-intent request
+    pub shipping: Option<Address>,
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/api_models/src/payments.rs:4488]: No, we should always accept `payment_method_data.billing` during confirm.

File: crates/hyperswitch_domain_models/src/payments.rs [hunk 6]
```diff
@@ -393,3 +401,16 @@ where
...
+#[cfg(feature = "v2")]
+#[derive(Clone)]
+pub struct PaymentConfirmData<F>
+where
+    F: Clone,
+{
+    pub flow: PhantomData<F>,
+    pub payment_intent: PaymentIntent,
+    pub payment_attempt: PaymentAttempt,
+    pub payment_method_data: Option<crate::payment_method_data::PaymentMethodData>,
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/hyperswitch_domain_models/src/payments.rs:453]: This is not the same payment method data that is stored in payment attempt. This is the domain model which stores the actual payment method data. The one in payment_attempt stores insensitive data

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 7]
```diff
@@ -2130,6 +2096,194 @@ async fn update_payment_method_status_and_ntid<F: Clone>(
...
+{
+    async fn update_tracker<'b>(
+        &'b self,
+        state: &'b SessionState,
+        mut payment_data: PaymentConfirmData<F>,
+        response: types::RouterData<F, types::PaymentsAuthorizeData, types::PaymentsResponseData>,
+        key_store: &domain::MerchantKeyStore,
+        storage_scheme: enums::MerchantStorageScheme,
+        locale: &Option<String>,
+        #[cfg(all(feature = "v1", feature = "dynamic_routing"))] routable_connector: Vec<
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_response.rs:2136]: V1 and V2 are mutually exclusive features, so v1 related field should be removed

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 8]
```diff
@@ -2130,6 +2096,194 @@ async fn update_payment_method_status_and_ntid<F: Clone>(
...
+            RoutableConnectorChoice,
+        >,
+        #[cfg(all(feature = "v1", feature = "dynamic_routing"))] business_profile: &domain::Profile,
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_response.rs:2139]: V1 and V2 are mutually exclusive features, so v1 related field should be removed

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/operations/payment_confirm_intent.rs [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            | common_enums::IntentStatus::Cancelled
+            | common_enums::IntentStatus::Processing
+            | common_enums::IntentStatus::RequiresCustomerAction
+            | common_enums::IntentStatus::RequiresMerchantAction
+            | common_enums::IntentStatus::RequiresCapture
+            | common_enums::IntentStatus::PartiallyCaptured
+            | common_enums::IntentStatus::RequiresConfirmation
+            | common_enums::IntentStatus::PartiallyCapturedAndCapturable => {
+                Err(errors::ApiErrorResponse::PaymentUnexpectedState {
+                    current_flow: "cofirm_intent".to_string(),
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm_intent.rs:59]: fix typos

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/core/payments/transformers.rs [hunk 10]
```diff
@@ -140,19 +163,212 @@ where
...
+        off_session: None,
+        setup_mandate_details: None,
+        confirm: true,
+        statement_descriptor_suffix: None,
+        statement_descriptor: None,
+        capture_method: Some(payment_data.payment_intent.capture_method),
+        amount: payment_data
+            .payment_attempt
+            .amount_details
+            .net_amount
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/transformers.rs:250]: yes, will take this up separately. Will need to make these as private fields as well and provide access through getters and setters

**Review by jarnura — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by ShankarSinghC — APPROVED** [review]

File: crates/diesel_models/src/user/sample_data.rs [hunk 11]
```diff
@@ -15,122 +15,120 @@ use crate::{
...
+//     // Used to verify compatibility with PaymentAttemptTable
+//     fn convert_into_normal_attempt_insert(self) -> PaymentAttemptNew {
+//         // PaymentAttemptNew {
+//         //     payment_id: self.payment_id,
+//         //     merchant_id: self.merchant_id,
+//         //     status: self.status,
+//         //     error_message: self.error_message,
+//         //     surcharge_amount: self.surcharge_amount,
+//         //     tax_amount: self.tax_amount,
+//         //     payment_method_id: self.payment_method_id,
```
Comment (apoorvdixit88) [inline crates/diesel_models/src/user/sample_data.rs:81]: having two set of comment `//` is redundant?

**Review by apoorvdixit88 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**